tools/stacksnoop.lua

```diff
@@ -41,52 +54,58 @@ return function(BPF, utils)
   local filter = ""
 
   if args.pid then
-    filter = [[
-      u32 pid;
-      pid = bpf_get_current_pid_tgid();
-      if (pid != %d) { return; }
-    ]] % args.pid
+    filter = "if (pid != %d) { return; }" % args.pid
   end
 
   local text = program:gsub("FILTER", filter)
   local bpf = BPF:new{text=text}
   bpf:attach_kprobe{event=args.fn, fn_name="trace_stack"}
 
   if BPF.num_open_kprobes() == 0 then
-    print("Function \"%s\" not found. Exiting." % args.fn)
+    print("Function \"%s\" not found. Exiting." % {args.fn})
     return
   end
 
   if args.verbose then
     print("%-18s %-12s %-6s %-3s %s" %
-        {"TIME(s)", "COMM", "PID", "CPU", "SYSCALL"})
+        {"TIME(s)", "COMM", "PID", "CPU", "FUNCTION"})
   else
-    print("%-18s %s" % {"TIME(s)", "SYSCALL"})
+    print("%-18s %s" % {"TIME(s)", "FUNCTION"})
   end
 
   local stack_traces = bpf:get_table("stack_traces")
-  local pipe = bpf:pipe()
+  local start_ts = utils.posix.time_ns()
 
-  while true do
-    local task, pid, cpu, flags, ts, msg = pipe:trace_fields()
-    local stack_id = string.match(msg, "stack_id=(%d+)")
+  local function print_event(cpu, event)
+    local ts = (utils.posix.time_ns() - start_ts) / 1e9
 
-    if stack_id then
-      if args.verbose then
-        print("%-18.9f %-12.12s %-6d %-3d %s" % {ts, task, pid, cpu, args.fn})
-      else
-        print("%-18.9f %s" % {ts, args.fn})
-      end
+    if args.verbose then
+      print("%-18.9f %-12.12s %-6d %-3d %s" %
+          {ts, ffi.string(event.comm), event.pid, cpu, args.fn})
+    else
+      print("%-18.9f %s" % {ts, args.fn})
+    end
 
-      for addr in stack_traces:walk(tonumber(stack_id)) do
-        local sym, offset = ksym:resolve(addr)
-        if args.offset then
-          print("\t%-16p %s+0x%x" % {addr, sym, tonumber(offset)})
-        else
-          print("\t%-16p %s" % {addr, sym})
-        end
+    for addr in stack_traces:walk(tonumber(event.stack_id)) do
+      local sym, offset = ksym:resolve(addr)
+      if args.offset then
+        print("\t%-16p %s+0x%x" % {addr, sym, tonumber(offset)})
+      else
+        print("\t%-16p %s" % {addr, sym})
       end
     end
+
     print()
   end
+
+  local TASK_COMM_LEN = 16 -- linux/sched.h
+
+  bpf:get_table("events"):open_perf_buffer(print_event, [[
+    struct {
+      uint64_t stack_id;
+      uint32_t pid;
+      char comm[%d];
+    }
+  ]] % {TASK_COMM_LEN})
```

Comment:
imo the c->lua binding are less verbose and cleaner than c->py. Just curious luajit dislikes 'char comm[TASK_COMM_LEN]' because the macro is undefined? any other way to pass it down to lua other than this '%' approach?
overall LGTM. thank you for fixing it all.